sound/soc/sof/sof-audio.c

```diff
@@ -495,23 +485,179 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
 
 	/* register machine driver, pass machine info as pdata */
 	audio_data->pdev_mach =
-		platform_device_register_data(sdev->dev, drv_name,
+		platform_device_register_data(audio_data->dev, drv_name,
 					      PLATFORM_DEVID_NONE, mach, size);
 	if (IS_ERR(audio_data->pdev_mach))
 		return PTR_ERR(audio_data->pdev_mach);
 
-	dev_dbg(sdev->dev, "created machine %s\n",
+	dev_dbg(audio_data->dev, "created machine %s\n",
 		dev_name(&audio_data->pdev_mach->dev));
 
 	return 0;
 }
 EXPORT_SYMBOL(sof_machine_register);
 
-void sof_machine_unregister(struct snd_sof_dev *sdev, void *pdata)
+void sof_machine_unregister(void *data)
 {
-	struct snd_sof_audio_data *audio_data = sdev->sof_audio_data;
+	struct snd_sof_audio_data *audio_data =
+		(struct snd_sof_audio_data *)data;
 
 	if (!IS_ERR_OR_NULL(audio_data->pdev_mach))
 		platform_device_unregister(audio_data->pdev_mach);
 }
 EXPORT_SYMBOL(sof_machine_unregister);
+
+static int sof_destroy_pipelines(struct device *dev)
+{
+	struct snd_sof_audio_data *audio_data =	sof_get_client_data(dev);
+	struct snd_sof_widget *swidget;
+	struct sof_ipc_free ipc_free;
+	struct sof_ipc_reply reply;
+	int ret = 0;
+
+	list_for_each_entry_reverse(swidget, &audio_data->widget_list, list) {
+		memset(&ipc_free, 0, sizeof(ipc_free));
+
+		/* skip if there is no private data */
+		if (!swidget->private)
+			continue;
+
+		/* configure ipc free message */
+		ipc_free.hdr.size = sizeof(ipc_free);
+		ipc_free.hdr.cmd = SOF_IPC_GLB_TPLG_MSG;
+		ipc_free.id = swidget->comp_id;
+
+		switch (swidget->id) {
+		case snd_soc_dapm_scheduler:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_PIPE_FREE;
+			break;
+		case snd_soc_dapm_buffer:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_BUFFER_FREE;
+			break;
+		default:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_COMP_FREE;
+			break;
+		}
+		ret = sof_client_ipc_tx_message(dev, ipc_free.hdr.cmd,
+						&ipc_free, sizeof(ipc_free),
+						&reply, sizeof(reply));
+		if (ret < 0) {
+			dev_err(dev,
+				"error: failed to free widget type %d with ID: %d\n",
+				swidget->widget->id, swidget->comp_id);
+
+			return ret;
+		}
```

Comment:
We probably want to keep going here and free as many as we can. Lets just log any failures.

Reply:
@lgirdwood that would probably work if we were to power down the DSP irrespective of the errors during free. But the DSP could be active because of another client and if we do not report the error here during suspend, a resume for this audio client will fail as well.

Reply:
I'm thinking here about dependencies or races. i.e. we are draining or we are a component that has dependencies (and cant be freed until all dependent components are freed). 

I the above 2 cases I would expect our components to return -EAGAIN (waiting for draining completion or other event completion) or -EBUSY (for dependencies). So far we have zero components that return these codes, but we should implement and document this in our APIs so that there is no future API change.

Reply:
@lgirdwood we free the pipeline components only in the case of runtime suspend. This should rule out both the above cases because there should be nothing active at all before runtime suspend kicks in. Does that make sense?

Reply:
We have the case when we may be forced to tear down a pipelines (say when a mediator guest crashes).
Some components also may need time to clear any HW state (think DMA buffer draining so may not be immediately ready to free). There is no harm in adding this exception now so that it's there when needed.